component/loki/process/internal/stages/geoip.go

```diff
@@ -51,49 +54,76 @@ var fields = map[GeoIPFields]string{
 
 // GeoIPConfig represents GeoIP stage config
 type GeoIPConfig struct {
-	DB     string  `river:"db,attr"`
-	Source *string `river:"source,attr"`
-	DBType string  `river:"db_type,attr"`
+	DB            string            `river:"db,attr"`
+	Source        *string           `river:"source,attr"`
+	DBType        string            `river:"db_type,attr,optional"`
+	CustomLookups map[string]string `river:"custom_lookups,attr,optional"`
 }
 
-func validateGeoIPConfig(c GeoIPConfig) error {
+func validateGeoIPConfig(c GeoIPConfig) (map[string]*jmespath.JMESPath, error) {
 	if c.DB == "" {
-		return ErrEmptyDBPathGeoIPStageConfig
+		return nil, ErrEmptyDBPathGeoIPStageConfig
 	}
-
 	if c.Source != nil && *c.Source == "" {
-		return ErrEmptySourceGeoIPStageConfig
+		return nil, ErrEmptySourceGeoIPStageConfig
+	}
+
+	if c.DBType == "" && c.CustomLookups == nil {
+		return nil, ErrEmptyDBTypeAndValuesGeoIPStageConfig
+	}
+
+	switch c.DBType {
+	case "", "asn", "city":
+	default:
+		return nil, ErrEmptyDBTypeGeoIPStageConfig
 	}
 
-	if c.DBType == "" {
-		return ErrEmptyDBTypeGeoIPStageConfig
+	if c.CustomLookups != nil {
```

Comment:
Is this check actually needed due to line 76+71? And if it is can it be invested so the code isn't as indented?

```
if c.CustomLookups == nil {
   return nil,nil
}
```

Reply:
71: I require if either `DBType` or `CustomLookups` is set
76: Is a validation of `DBType` value

I'll fix so that the the indentation is removed